Improve sidebar style for deeply nested pages

thesuperzapper opened the pull request:
The current sidebar is almost unusable for the deeply nested pages (like those in KFP). The sidebar gets all bunched up and it's hard to know which pages are in what folders.

This PR significantly improves the readability of the sidebar links by doing the following:

1. Expands the sidebar width
2. Remove unnecessary padding and indents
3. Putting a vertical line to show where each folder starts and ends
4. Bolding the "chain" of pages which leads to the currently active page

This PR also does the following:

- applies a max-width to the site, to prevent very wide screens making the site hard to read
- fixes the top navigation bar for smaller screens, the issue was the long "event promo" links were ending up underneath the logo
- adds a subtle rotation on the "expand section" arrow when you are hovering
- removes an invalid `OWNERS.md` file (note the `.md`), which was showing up as a nameless page.

# Laptop Screenshots

## OLD

![laptop_old](https://github.com/user-attachments/assets/21107daf-54ba-4158-9cd2-9e5091334829)

## NEW

![laptop_new](https://github.com/user-attachments/assets/22b3816e-4b67-4c91-ac2a-879c879d7d6b)

# Widescreen Screenshots

## OLD

![widescreen_old](https://github.com/user-attachments/assets/c004ce92-025e-45a1-8b7f-b379ebfbe2a9)

## NEW

![widescreen_new](https://github.com/user-attachments/assets/f0a078ec-c6b1-4ba3-8f29-34a905cb11a5)

# Mobile

## NEW <--> OLD

<img src="https://github.com/user-attachments/assets/c56d4082-759f-437b-880c-ca4926d977a3" width="40%">
<-->
<img src="https://github.com/user-attachments/assets/106eebe7-5daa-455a-a6a9-7f1e501ba051" width="40%">

thesuperzapper commented:
@james-jwu @jbottum @andreyvelich @terrytangyuan can one of you please review or approve this website PR?

It's a very important readability change, and makes no changes to the content so it should not be controversial.

andreyvelich commented:
Thank you for this. Additionally, we should discuss how we can decrease the left panel since this change will reduce the actual size of the document (cc @StefanoFioravanzo )


<img width="677" alt="Screenshot 2024-07-22 at 16 45 03" src="https://github.com/user-attachments/assets/d82ba279-6bc3-4061-9c01-1e207f42e07d">


<img width="765" alt="Screenshot 2024-07-22 at 16 48 51" src="https://github.com/user-attachments/assets/39d0e6d6-d7ac-4822-a3f7-17ffcddcef95">

StefanoFioravanzo commented:
@thesuperzapper Thanks for this, it's a great UX improvement.

I back @ederign's proposal - can we remove the ~30 degree on-hover animation? I get why you put it, it can be a nice visual touch, but I feel like it overstimulates in this case. 

@andreyvelich  On the sidebar width, we do have an issue with a content tree that's too deep. When the sidebar is too narrow, you end up with folding titles. In the current preview the sidebar does seem too wide though. @thesuperzapper would you consider finding a balance between the current width and the one you propose? Maybe something in the middle would work?

Overall awesome change, thanks

ederign commented:
Locally, the left panel sidebar issue only happens on large screens. In the end, it is a matter of taste. If we:

If we change the left sidebar to col-xl-2 (instead of col-xl-3) and the main to col-xl-8 (from col-xl-7), it will look like this:

<img width="2558" alt="image" src="https://github.com/user-attachments/assets/8d198010-3071-4bff-a4c4-ad9451114f98">

Also, for some reason (I don't have the context), the content was to max-width:80%, which makes the main use only a partial size of the screen. On the screenshot, I've changed to max-width:100% locally, but not sure if this breaks anything.

thesuperzapper commented:
@ederign @andreyvelich @StefanoFioravanzo thanks for everyone's feedback.

I have made the following changes:

1. removed the rotation on the hovering of the arrows
2. made the sidebar slightly smaller
3. fixed the issue raised by @ederign about content only being able to use 80% of the main section
4. applied a max-width to make the site much more readable on very wide screens:
     - _This was a nightmare to get working cleanly across the entire site, but I think it has significantly improved readability for wide screens, try it for yourself and see._

Please take another look on [the preview site](https://deploy-preview-3816--competent-brattain-de2d6d.netlify.app/docs/started/) but I think this PR should be ready to merge now.

ederign commented:
1 and 2 are fixed!

For 3 and 4, I found an issue:

![image](https://github.com/user-attachments/assets/2112145b-c63b-4bf8-beb4-3d646ef9e1f8)

andreyvelich commented:
After this change I can see that content is centralized on the big screens and only occupies 50%:
<img width="2545" alt="Screenshot 2024-07-23 at 13 53 40" src="https://github.com/user-attachments/assets/6840c5f3-6c4d-42f0-b4b0-ec9d835b044c">
<img width="2541" alt="Screenshot 2024-07-23 at 13 53 46" src="https://github.com/user-attachments/assets/a58b1754-f648-4de8-a1f8-3f3e0c589691">

StefanoFioravanzo commented:
Suggestion that may be out of scope for this PR: How about we auto-collapse the current open top-level menu when you open a new one? So we avoid an excessively long sidebar tree

![Screenshot 2024-07-23 at 15 16 47](https://github.com/user-attachments/assets/49b166f0-8b85-40c2-bf2d-5ebfb28e8e32)

thesuperzapper commented:
@ederign that issue was actually present in the existing page, it was caused by someone including the entirety of bootstrap CSS in a <style> tag (because the page is auto generated by swagger).

I have manually removed that bootstrap code and it mostly fixed the style issues on that specific page (it's not perfect, but it's still better than before).

thesuperzapper commented:
@StefanoFioravanzo this is already the behavior, when you click on a page, only the "chain" leading to your current page remains open.

It would be very unexpected for us to close pages while people were "browsing" for a page (before clicking).

thesuperzapper commented:
> After this change I can see that content is centralized on the big screens and only occupies 50%

@andreyvelich is this a positive or negative comment? 

The width-limit increases the readability by preventing very wide screens from having content on the far edges of the screen.

If there are no more comments, and people are ok with the width-limit, I think we can merge this.

thesuperzapper commented:
I did one last small cleanup to the styling in https://github.com/kubeflow/website/pull/3816/commits/1f89a55465360a1d5479162a49a471a13ca8d7c9.

The main fixes were:

1. Homepage is now properly centered
2. On small screens, there was a slight stripe of the darker border color, which should have only been visible on larger screens (with sidebars)
3. Added a small border outline at the top of the page on mobile (below the search box).

I think this should be ready to merge now.

andreyvelich commented:
> > After this change I can see that content is centralized on the big screens and only occupies 50%
> 
> @andreyvelich is this a positive or negative comment?
> 
> The width-limit increases the readability by preventing very wide screens from having content on the far edges of the screen.
> 
> If there are no more comments, and people are ok with the width-limit, I think we can merge this.

From my point of view this is makes our docs harder to read since the documentation area is very small.
Can we just follow the same settings as Kubernetes docs since they also use Hugo ?
https://kubernetes.io/docs/concepts/overview/components/

I prefer their settings.

thesuperzapper commented:
> @thesuperzapper 
> 
> > This is already the behavior, when you click on a page, only the "chain" leading to your current page remains open.
> 
> That's not true. See my screenshot above. If what you said was true, how could I have found myself with that content tree?

@StefanoFioravanzo once you __click__ on any page everything but the current page will close. Note, I am not talking about clicking the arrow, I am talking about clicking the page.

Similarly when you refresh the page all other open sections are closed.

Either way, feel free to raise a separate PR if you want different behavior.

thesuperzapper commented:
> > > After this change I can see that content is centralized on the big screens and only occupies 50%
> > 
> > @andreyvelich is this a positive or negative comment?
> > 
> > The width-limit increases the readability by preventing very wide screens from having content on the far edges of the screen.
> > 
> > If there are no more comments, and people are ok with the width-limit, I think we can merge this.
> 
> From my point of view this makes our docs harder to read since the documentation area is very small.
> Can we just follow the same settings as Kubernetes docs since they also use Hugo ?
> https://kubernetes.io/docs/concepts/overview/components/
> 
> I prefer their settings.

@andreyvelich having a max-width makes reading the page easier. Without it, the page is hard to read on a very wide screen.

It's well established that people find it easier to read things with margins and shorter line lengths.  Give it a try yourself, show people the two pages and ask which they prefer.

If you are concerned about the specific max-width I have used, this is easy to change with a single variable. However, in my testing (and when compared to other websites), a max-width of 1440px/100rem was typical and easiest to read.

For reference, other docs websites already do this this like ArgoCD/Workflows, Istio and more.

_PS: what's also interesting is that even with the width-limit, the height of the page remains the same in most cases._

_PSS: let's do some user testing, or just ask a few other community members. But remember, this change affects only a small fraction of users, as almost no one reads these pages on a wide/4k screen._

andreyvelich commented:
>It's well established that people find it easier to read things with margins and shorter line lengths. Give it a try yourself, show people the two pages and ask which they prefer.

I mean this is very subjective, we can't say it for all of the users. 

Looking at other Kubernetes project, they all have wider section for doc content:
- Kubernetes: https://kubernetes.io/docs/tutorials/hello-minikube/
- Kueue: https://kueue.sigs.k8s.io/docs/concepts/admission_check/
- Istio: https://istio.io/latest/docs/overview/what-is-istio/


My feeling is that we should just make the same settings as Kubernetes Hugo website, given that their website has the largest user attention.

thesuperzapper commented:
> > It's well established that people find it easier to read things with margins and shorter line lengths. Give it a try yourself, show people the two pages and ask which they prefer.
> 
> I mean this is very subjective, we can't say it for all of the users.
> 
> Looking at other Kubernetes project, they all have wider section for doc content:
> 
> * Kubernetes: https://kubernetes.io/docs/tutorials/hello-minikube/
> * Kueue: https://kueue.sigs.k8s.io/docs/concepts/admission_check/
> * Istio: https://istio.io/latest/docs/overview/what-is-istio/
> 
> My feeling is that we should just make the same settings as Kubernetes Hugo website, given that their website has the largest user attention.

@andreyvelich the Istio site you linked uses a max-width of `1440px` (like what I am proposing in this PR), and I think it's much easier to read than the Kubernetes website.

If we are talking about the Kubernetes website specifically, it's likely not because they don't want a width-limit, but because the default theme does not have it, and it was quite complex to implement.

Here is a screenshot of the Istio website:

![istio_wide](https://github.com/user-attachments/assets/82548692-40ff-430d-b642-307087330f2f)

andreyvelich commented:
Yeah, but even Istio has large width (902 vs 768) for content.

![Screenshot 2024-07-24 at 22 39 00](https://github.com/user-attachments/assets/17f26f2a-ea0d-459c-a4cb-5b113d0bfd8d) 
![Screenshot 2024-07-24 at 22 38 48](https://github.com/user-attachments/assets/3d3b7e50-d3a0-4684-b156-4c4827a2924e)

thesuperzapper commented:
@andreyvelich as long as we agree a fixed-width makes sense, we can discuss making the sidebar smaller.

But I still think it's worth keeping the wider sidebar because we have much more deeply nested pages than istio.

Perhaps @ederign can share some of his design experience and weigh in on what he thinks is best?

---

BTW, I am going to be on vacation till next Monday 30th July, so if you need any follow up commits to this (or other) PRs, it will need to wait till then.

This PR should be ready to merge (if we all can agree on the sidebar width), so feel free to do that if you want these fixes in.

@ederign could make a follow up PR if needed, because I think the changes are worth having soon, with the influx of 1.9 users.

ederign commented:
I agree with @thesuperzapper that having a fixed max-width makes reading the page easier and more accessible. A max width of 1440px/100rem sounds to me like a good starting point for large screens, especially when balanced with the depth of our submenus.

If people are interested in reading more about this:

> Having the right amount of characters on each line is key to the readability of your text.
...
Too wide: if a line of text is too long the reader’s eyes will have a hard time focusing on the text. This is because the line length makes it difficult to gauge where the line starts and ends. Furthermore it can be difficult to continue onto the correct line in large blocks of text.
 [[link](https://baymard.com/blog/line-length-readability)]
 
![image](https://github.com/user-attachments/assets/a15d0b54-580a-4630-94e7-51dd75ab3bb6)

> If a font size is too small and the line length too long, readers will have to strain to read the text, which causes their eyes to fatigue [[link](https://kickpoint.ca/the-readability-formula-making-your-website-easy-to-read/)]


We can always discuss the size of our main panel, but I think a maximum width is important for readability. 

This PR looks good to me!

ederign commented:
/lgtm

andreyvelich commented:
From my point of view, we should at least follow the Istio settings.
Right now the area for docs looks very small.

thesuperzapper commented:
@andreyvelich in the interest of getting this PR merged, I have reverted the sidebar to be the same size as the current website.

I think we are good to merge if @andreyvelich and @ederign give it a look over.

---

Also, in the last commit I made the following changes:

1. Fixed a bug where the homepage header would disappear on large mobile screens
2. Implemented a proper swagger-ui page for the pipelines API:
      - [Preview: v1beta1](https://deploy-preview-3816--competent-brattain-de2d6d.netlify.app/docs/components/pipelines/legacy-v1/reference/api/kubeflow-pipeline-api-spec/)
      - [Preview: v2beta1](https://deploy-preview-3816--competent-brattain-de2d6d.netlify.app/docs/components/pipelines/reference/api/kubeflow-pipeline-api-spec/)
3. Made the "feedback on this page" links pre-populate the issue with which page the user was on

ederign commented on layouts/shortcodes/swaggerui-inline.html:
```diff
@@ -0,0 +1,46 @@
+
+    const ui = SwaggerUIBundle({
+      url: {{ $original | relURL }},
+      domNode: swaggerNode,
```
Suggestion to disable 
supportedSubmitMethods: []
"An empty array disables "Try it out" for all operations. This does not filter the operations from the display."

https://swagger.io/docs/open-source-tools/swagger-ui/usage/configuration/

ederign reviewed (changes requested):
@thesuperzapper :

- Fixed a bug where the homepage header would disappear on large mobile screens

I cannot reproduce it anymore.

- Implemented a proper swagger-ui page for the pipelines API:

I like it. My only suggestion would be to disable the tryout action. I've looked at the Swagger API and left the code there that should work. (I didn't try it.)

- Made the "feedback on this page" links pre-populate the issue with which page the user was on

Nice touch.

ederign commented:
This PR includes important changes that improve the web page, and I would like to see it merged as soon as possible. I'm totally fine with having the 'Try out' issue fixed in another PR (I can even send the PR myself later).

thesuperzapper commented:
@ederign rather than disabling the "try it out" feature, I have made a system to allow users to put their own URL in.

It will still not work unless they have the appropriate authentication setup, but it makes it much easier to use the UI to formulate a request to run locally.

I have also added a preamble that explains a bit more about the API and how to use it on the Swagger pages.

Please review again and LGTM if you are happy now.

ederign commented:
@thesuperzapper is this header expected? (at [link](https://deploy-preview-3816--competent-brattain-de2d6d.netlify.app/docs/components/pipelines/reference/api/kubeflow-pipeline-api-spec/))

<img width="821" alt="image" src="https://github.com/user-attachments/assets/a88cd0ab-bb48-4715-b56a-777a6a5baa75">

thesuperzapper commented:
> @thesuperzapper is this header expected? (at [link](https://deploy-preview-3816--competent-brattain-de2d6d.netlify.app/docs/components/pipelines/reference/api/kubeflow-pipeline-api-spec/))
> 
> <img alt="image" width="821" src="https://private-user-images.githubusercontent.com/531351/355085146-a88cd0ab-bb48-4715-b56a-777a6a5baa75.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI5NjAyMDcsIm5iZiI6MTcyMjk1OTkwNywicGF0aCI6Ii81MzEzNTEvMzU1MDg1MTQ2LWE4OGNkMGFiLWJiNDgtNDcxNS1iNTZhLTc3N2E2YTViYWE3NS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODA2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgwNlQxNTU4MjdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jOTA5YmU5ODk4MWNkNGM4NDZjNmVjNTQ1MTcxM2NmYmJhN2Y5MjdiZjJhYmVjOTU1YjgxNWU4YjY5YTIxYTk2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.FLgFPI2yPuWUtF6K90T5OcB9BOIwsM8EtVWfA4_CNOo">

@ederign I have changed the way the Swagger JSON is embedded so I can remove the hard-coded introduction/google-email, take a look:
- [v2 API Page](https://deploy-preview-3816--competent-brattain-de2d6d.netlify.app/docs/components/pipelines/reference/api/kubeflow-pipeline-api-spec/)
- [v1 API Page](https://deploy-preview-3816--competent-brattain-de2d6d.netlify.app/docs/components/pipelines/legacy-v1/reference/api/kubeflow-pipeline-api-spec/)

thesuperzapper commented:
@ederign I have made some final tweaks based on feedback from people I showed:

- made it so the logo is always the leftmost thing on the page (looks better)
- fixed the sidebar color bleeding into the page on mobile screens (where there are no sidebars)
- improved the padding of the table-of-contents on mobile (it was squished up on the left side of the screen)

People seem to be happy with it now, so if you can do a final review, then we can merge this.
For reference, here is a page: [OLD](https://www.kubeflow.org/docs/started/introduction/), [NEW](https://deploy-preview-3816--competent-brattain-de2d6d.netlify.app/docs/started/introduction/)

ederign commented:
@thesuperzapper LGTM to me. Thank you so much for accommodating our requests.

/lgtm

thesuperzapper commented:
@andreyvelich can you please approve if you are happy with the changes?

andreyvelich commented:
Thanks for the update @thesuperzapper.
@thesuperzapper @ederign Do we make the width for the content similar to Istio, 902px: https://github.com/kubeflow/website/pull/3816#issuecomment-2248949756 ?
Right now, it is 777px.

/assign @kubeflow/wg-pipeline-leads @rimolive
This PR has some changes into KFP docs to add Swagger for API server, please take a look.

thesuperzapper commented:
@andreyvelich on a 4k screen the content area is currently 864px (with padding it is 960px).

<img width="244" alt="Screenshot 2024-08-08 at 18 07 04" src="https://github.com/user-attachments/assets/c115233f-cfbd-4477-827b-cba7e0475014">

Personally, I think it's very readable right now, and making it any wider or decreasing the padding/margins would decrease the readability.

thesuperzapper commented:
For the @kubeflow/wg-pipeline-leads reviewers, here is a direct link to the new Swagger UI pages:

- [v2 REST API](https://deploy-preview-3816--competent-brattain-de2d6d.netlify.app/docs/components/pipelines/reference/api/kubeflow-pipeline-api-spec/)
- [v1 REST API](https://deploy-preview-3816--competent-brattain-de2d6d.netlify.app/docs/components/pipelines/legacy-v1/reference/api/kubeflow-pipeline-api-spec/)

andreyvelich commented:
Sure, maybe we can start with that setup.
/lgtm
cc @jbottum @kubeflow/wg-pipeline-leads @rimolive

rimolive commented:
> Thanks for the update @thesuperzapper. @thesuperzapper @ederign Do we make the width for the content similar to Istio, 902px: [#3816 (comment)](https://github.com/kubeflow/website/pull/3816#issuecomment-2248949756) ? Right now, it is 777px.
> 
> /assign @kubeflow/wg-pipeline-leads @rimolive This PR has some changes into KFP docs to add Swagger for API server, please take a look.

Changes in the pipelines docs look great to me.

/lgtm

terrytangyuan reviewed (approved):
/lgtm
/approve
